### PR TITLE
Several fixes, new formats, and tons of tests

### PR DESCRIPTION
**Pull request by hillmanov**
First off, thank you for this library! It has served our needs well for some time, but recently our needs have changed. We needed support for a LOT of different and odd date formats that were not originally caught by this library, so I made the changes necessary to support our needs. The vast majority of the tests were generated programmatically, which is why there are so many. 

During the process some of the open issues were addressed, so they are included in this PR as well. 

If the changes to the code or the tests are not inline with what you have in mind, that is just fine. My organization will continue to use our fork unless this gets merged in. 

Let me know how you would like to proceed!

### CODE REVIEW

**Comment by gr2m** [comment]
Wow thanks Scott! I know you put a lot time into this, but because it’s such a big PR, it will take lots of time for me to review, time I don’t have right now due to family. 

Can you please split it up? Otherwise I fear this will get blocked for a long time, it would be better to get some things merged we agree on and have separate PRs for things that we can discuss, without blocking everything.

**Comment by hillmanov** [comment]
@gr2m 

I completely understand - t is a very large PR! The vast majority of the changes are in the tests, with a pretty good handful of new regular expressions and `format = format.replace...` statements.

I am not worried personally about the timeline of this getting in, and if it takes a long time to get to that is just fine for me. 

As far as breaking it up, I know and understand that would be ideal, however, that would prove to be pretty difficult to do. I knew which formats I had to support coming in so I created almost all of the tests at that point (over 1,200 failed first run!) and I just churned until they all passed. A fix for one section would break another so in the end it felt like an 'all or nothing' kind of PR. This isn't how I like to develop software, but we needed support for these formats pretty fast so the rest of the project could continue. 

I will see what I have time for myself since now that it is working the way we need, I can move on with my project.

File: package.json [hunk 1]
```diff
@@ -60,5 +60,5 @@
       "demo"
     ]
   },
-  "version": "0.0.0-development"
+  "version": "1.0.0"
```
Comment (gr2m) [inline package.json]: `moment-parseformat` is published automatically using semantic-release, the new version is calculated automatically based on commit message conventions. Last release is 3.0.0: https://github.com/gr2m/moment-parseformat/releases

**Comment by RichardCr** [comment]
@gr2m Hi there! Any chance this can be merged in? This fixes several issues for me. (07 Dec 20 -> without this gets DD MMM D)

File: test/moment-parseformat-test.js [hunk 2]
```diff
@@ -42,7 +42,7 @@ test('GitHub issues', function (t) {
   t.equal(moment.parseFormat('01-01-2015'), 'DD-MM-YYYY', '#15 01-01-2015 → DD-MM-YYYY')
   t.equal(moment.parseFormat('01-01-2015', { preferredOrder: 'MDY' }), 'MM-DD-YYYY', '#15 01-01-2015 (preferredOrder: MDY) → MM-DD-YYYY')
 
-  // https://github.com/gr2m/moment.parseFormat/pull/21
+  // // https://github.com/gr2m/moment.parseFormat/pull/21
```
Comment (gr2m) [inline test/moment-parseformat-test.js:45]: can you revert the comment changes

File: test/moment-parseformat-test.js [hunk 3]
```diff
@@ -113,3 +112,3820 @@ test('Github issue #40', function (t) {
...
+})
+
+test('Github issue #60', function (t) {
+  t.equal(moment.parseFormat('05 2015'), 'MM YYYY', '05 2015 → MM YYYY')
+  t.end()
+})
+
+// Partially. Need to pass in a flag to get the DD, since this single D format is favored everywhere else
+test('Github issue #61', function (t) {
+  t.equal(moment.parseFormat('19-feb-1990', { preferLongFormat: true }), 'DD-MMM-YYYY', '05 2015 → DD-MMM-YYYY')
```
Comment (gr2m) [inline test/moment-parseformat-test.js:133]:
```suggestion
  t.equal(moment.parseFormat('19-feb-1990', { preferLongFormat: true }), 'DD-MMM-YYYY', '19-feb-1990 → DD-MMM-YYYY')
```

**Review by gr2m — COMMENTED** [review]
the PR looks great overall. I was reluctant because it adds so much logic, but I think the value is worth the costs. Can you address the remaining comments please?

**Comment by gr2m** [comment]
@RichardCr I've reviewed the PR in full now. I can go ahead and merge it and then do follow up comments to address my comments, but I want to give @hillmanov a week to address them first

**Comment by hillmanov** [comment]
@gr2m and @RichardCr I'll see what I can do. I no longer work at the company I was at when I made this PR, so I don't have the ability to push stuff to that forked repo anymore. However, I am still close with several of the devs there, so I am hopeful that I can get something done.

**Comment by RichardCr** [comment]
@gr2m Hi there, just seeing if it is likely this will get merged or not. Thanks!

File: test/moment-parseformat-test.js (hunk 2 above)
Comment (gr2m) [reply test/moment-parseformat-test.js:45]:
```suggestion
  // https://github.com/gr2m/moment.parseFormat/pull/21
```

**Review by gr2m — APPROVED** [review]

**Comment by gr2m** [comment]
> @gr2m Hi there, just seeing if it is likely this will get merged or not. Thanks!

can you help resolve the conflicts?

**Comment by hillmanov** [comment]
@gr2m conflicts resolved (finally) and all tests pass. I updated many of the test cases to get them compliant with the standard formatting requirement as well.

Not sure what is going on with the failing travis-ci build though - Node 6 is just failing to install some things, but it doesn't look like it has anything to do with this PR. 

I also see that the code coverage percent went down, but this PR also increases the amount of test by... a lot.

**Comment by RichardCr** [comment]
> > @gr2m Hi there, just seeing if it is likely this will get merged or not. Thanks!
> 
> can you help resolve the conflicts?

I'm still pretty new to all this, I wouldn't rate myself not to break something. Happy to try though.

**Comment by hillmanov** [comment]
@RichardCr  I was finally able to clean my PR up as far as the conflicts are concerned. There looks to be some kind of issue with the CI checks with Node 6 - something isn't running successfully, but it doesn't look like it is related to this PR. I think we just need to wait to hear from @gr2m now.

**Comment by RichardCr** [comment]
@hillmanov Thanks so much. I really appreciate everyone who has contributed to this!

**Review by gr2m — APPROVED** [review]
Thank you everyone!

**Comment by gr2m** [comment]
It looks like we accidentally introduced a breaking change: https://github.com/gr2m/moment-parseformat/issues/109.

I'll revert these changes and publish 3.1.1, then release a breaking version 4.0.0 with the changes here

**Comment by gr2m** [comment]
now released as v4.0.0

**Pull request merged**